src/access/wac.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { internal_fetchAcl } from "../acl/acl.internal";
+import { getAgentAccess as getAgentAccessWac } from "../acl/agent";
+import { WebId, WithServerResourceInfo } from "../interfaces";
+import { internal_defaultFetchOptions } from "../resource/resource";
+import { Access } from "./universal";
+import { Access as AclAccess } from "../acl/acl";
+
+// Setting WacAccess = Required<Access> enforces that any change on Access will
+// reflect on WacAccess, but WacAccess has additional restrictions.
+type WacAccess = Required<Access> &
+  (
+    | { controlRead: true; controlWrite: true }
+    | ({ controlRead: undefined; controlWrite: undefined } & {
+        read: true | undefined;
+        append: true | undefined;
+        write: true | undefined;
+      })
```

Comment:
I think the brackets here might be the wrong way around? I'd think it'd be something like `(<control true> | <control undefined>) & <read append write>` (but I might be getting my De Morgan's mixed up)

(Yes I looked "De Morgan's laws" up to sound clever.)

Reply:
Well it worked, because I looked it up as well, and also the brackets were messed up. 90ec90d4c1de96c4ef16c609e6047a7c951a8cb6